Update .NET intro (#20811)
dotnet/docs: docs/core/introduction.md

With .NET, your code and project files look and feel the same no matter which type of app you're building. You have access to the same runtime, API, and language capabilities with each app.

## Cross platform

You can create .NET apps for many operating systems, including:

* Windows
* macOS
* Linux
* Android
* iOS
* tvOS
* watchOS

Supported processor architectures include:

* x64
* x86
* ARM32
* ARM64

.NET lets you use platform-specific capabilities, such as operating system APIs. Examples are Windows Forms and WPF on Windows and the native bindings to each mobile platform from Xamarin.

For more information, see [Supported OS lifecycle policy](https://github.com/dotnet/core/blob/master/os-lifecycle-policy.md) and [.NET RID Catalog](rid-catalog.md).

## Open source

.NET is open source, using [MIT and Apache 2 licenses](https://github.com/dotnet/runtime/blob/master/LICENSE.TXT). .NET is a project of the [.NET Foundation](https://dotnetfoundation.org/).

For more information, see the [list of project repositories on GitHub.com](https://github.com/dotnet/core/blob/master/Documentation/core-repos.md).

## Support

> .NET is supported by Microsoft on Windows, macOS, and Linux. It's updated for security and quality regularly, on the second Tuesday of each month.

Reviewer: ```suggestion
.NET is supported by Microsoft on Windows, macOS, and Linux. It's updated regularly for security and quality, on the second Tuesday of each month.
```

* [Generic types](../standard/generics.md)
* [Delegates](../standard/delegates-lambdas.md)
* [Lambdas](../standard/delegates-lambdas.md)
* [Events](../standard/events/index.md)
* [Exceptions](../standard/exceptions/index.md)
* [Attributes](../standard/attributes/index.md)
* [Asynchronous code](../standard/async.md)
* [Parallel programming](../standard/parallel-programming/index.md)
* [Code analyzers](../fundamentals/code-analysis/overview.md)

### IDEs

The integrated development environments for .NET include:

* [Visual Studio](https://visualstudio.microsoft.com/vs/)

  Runs on Windows only. Has extensive built-in functionality designed to work with .NET. The Community edition is free for students, open-source contributors, and individuals.

* [Visual Studio Code](https://code.visualstudio.com/)

  Runs on Windows, macOS, and Linux. Free and open source. Extensions are available for working with .NET languages.

* [Visual Studio for Mac](https://visualstudio.microsoft.com/vs/mac/)

>   For developing .NET apps and games for iOS, Android, and web.

Reviewer: Do we want to say "Runs on macOS only" for clarity?

* [GitHub Codespaces](https://github.com/features/codespaces)

  An online Visual Studio Code environment, currently in beta.

### SDK and runtimes

The [.NET SDK](sdk.md) is a set of libraries and tools for developing and running .NET applications.

> When you [download .NET](https://dotnet.microsoft.com/download/dotnet-core/), you can choose the SDK or a *runtime*. Install a runtime on a machine that you want to prepare for running .NET apps. Install the SDK on a machine that you want to use for development. When you download the SDK, you automatically get the runtimes with it.

Reviewer: ```suggestion
When you [download .NET](https://dotnet.microsoft.com/download/dotnet-core/), you can choose the SDK or a *runtime*, such as the ASP.NET Core runtime or the desktop runtime. Install a runtime on a machine that you want to prepare for running .NET apps. Install the SDK on a machine that you want to use for development. When you download the SDK, you automatically get the runtimes with it.
```

* [Using .NET notebooks with Jupyter on your machine](https://github.com/dotnet/interactive/blob/main/docs/NotebooksLocalExperience.md)
* [.NET Interactive documentation](https://github.com/dotnet/interactive/blob/main/docs/README.md)

## Execution models

.NET apps run [managed code](../standard/managed-code.md) in a runtime environment known as the Common Language Runtime (CLR).

### CLR

The .NET [CLR](../standard/clr.md) is a cross-platform runtime that includes support for Windows, macOS, and Linux. The CLR handles memory allocation and management. The CLR is also a virtual machine that not only executes apps but also generates and compiles code using a just-in-time (JIT) compiler.

For more information, see [Common Language Runtime (CLR) overview](../standard/clr.md).

### JIT compiler and IL

Higher-level .NET languages, such as C#, compile down to a hardware-agnostic instruction set, which is called Intermediate Language (IL). When an app runs, the JIT compiler translates IL to machine code that the processor understands. JIT compilation happens on the same machine that the code is going to run on.

> Since JIT compilation occurs during execution of the application, compile time is part of the run time. Therefore, JIT compilers have to balance time spent optimizing code against the savings that the resulting code can produce. But a JIT compiler knows the actual hardware and can free developers from having to ship different implementations for different platforms.

Reviewer: ```suggestion
Since JIT compilation occurs during execution of the application, the compile time is part of the run time. Therefore, JIT compilers have to balance time spent optimizing code against the savings that the resulting code can produce. But a JIT compiler knows the actual hardware and can free developers from having to ship different implementations for different platforms.
```

The .NET JIT compiler can do *tiered compilation*, which means it can recompile individual methods at run time. This feature lets it compile quickly while still being able to produce a highly tuned version of the code for frequently used methods.

For more information, see [Managed execution process](../standard/managed-execution-process.md) and [Tiered compilation](whats-new/dotnet-core-3-0.md#tiered-compilation).

### AOT compiler

> The default experience for most .NET workloads is the JIT-compiler, but .NET offers two forms of ahead-of-time (AOT) compilation:

Reviewer: ```suggestion
The default experience for most .NET workloads is the JIT compiler, but .NET offers two forms of ahead-of-time (AOT) compilation:
```

* Primitive types, such as <xref:System.Boolean?displayProperty=nameWithType> and <xref:System.Int32?displayProperty=nameWithType>.
* Collections, such as <xref:System.Collections.Generic.List%601?displayProperty=nameWithType> and <xref:System.Collections.Generic.Dictionary%602?displayProperty=nameWithType>.
* Data types, such as <xref:System.Data.DataSet?displayProperty=nameWithType> and <xref:System.Data.DataTable?displayProperty=nameWithType>.
* Network utility types, such as <xref:System.Net.Http.HttpClient?displayProperty=nameWithType>.
* [File and stream I/O](../standard/io/index.md) utility types, such as <xref:System.IO.FileStream?displayProperty=nameWithType> and <xref:System.IO.TextWriter?displayProperty=nameWithType>.
* [Serialization](../standard/serialization/index.md) utility types, such as <xref:System.Text.Json.JsonSerializer?displayProperty=nameWithType> and <xref:System.Xml.Serialization.XmlSerializer?displayProperty=nameWithType>.
* High-performance types, such as <xref:System.Span%601?displayProperty=nameWithType>, <xref:System.Numerics.Vector?displayProperty=nameWithType>, and [Pipelines](../standard/io/pipelines.md).

> For more information, see [Framework libraries](../standard/framework-libraries.md) and [the source code for the libraries](https://github.com/dotnet/runtime/tree/master/src/libraries).

Reviewer: Do we want to link to the API browser?

Author: I think the number of framework choices on the API browser home page would be confusing for people at this point.

## Microsoft.Extensions libraries

Libraries for some commonly used application functionality aren't included in the runtime libraries but are made available in NuGet packages, such as the following:

| NuGet package | Documentation |
|---------|---------|
| [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) | [Application lifetime management (Generic Host)](extensions/generic-host.md) |
| [Microsoft.Extensions.DependencyInjection](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection) | [Dependency injection (DI)](extensions/dependency-injection.md)
| [Microsoft.Extensions.Configuration](https://www.nuget.org/packages/Microsoft.Extensions.Configuration) | [Configuration](extensions/configuration.md) |
| [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging) | [Logging](/aspnet/core/fundamentals/logging) |

For more information, see the [dotnet/extensions repository on GitHub](https://github.com/dotnet/extensions).

## Data access

.NET provides an Object/Relational Mapper (ORM) and a way to write SQL queries in code.

### Entity Framework Core

> Entity Framework (EF) Core is an [open source](https://github.com/aspnet/EntityFrameworkCore) and cross platform data access technology that can serve as an ORM. EF Core lets you work with a database by referring to .NET objects in code. It reduces the amount of data-access code you would otherwise need to write and test. EF Core supports many database engines.

Reviewer: ```suggestion
Entity Framework (EF) Core is an [open source](https://github.com/aspnet/EntityFrameworkCore) and cross-platform data-access technology that can serve as an ORM. EF Core lets you work with a database by referring to .NET objects in code. It reduces the amount of data-access code you would otherwise need to write and test. EF Core supports many database engines.
```

For more information, see [Entity Framework Core](/ef/core/) and [Database Providers](/ef/core/providers/).

### LINQ

Language-integrated query (LINQ) lets you write declarative code for operating on data. The data can be in many forms (such as in-memory objects, a SQL database, or an XML document), but the LINQ code you write typically doesn't differ by data source.

For more information, see [LINQ (Language Integrated Query) overview](../standard/linq/index.md).

## .NET terminology

To understand .NET documentation, it can help to know how the usage of some terms has changed over time.

### .NET Core and .NET 5

In 2002, Microsoft released [.NET Framework](../framework/get-started/overview.md), a development platform for creating Windows apps. Today .NET Framework is at version 4.8 and is still [supported by Microsoft](https://dotnet.microsoft.com/platform/support/policy/dotnet-framework).

> In 2014, Microsoft began writing a cross platform open source successor to .NET Framework. This new implementation of .NET was named .NET Core until it reached version 3.1. The next version after .NET Core 3.1 is .NET 5, currently in preview. Version number 4 was skipped to avoid confusion between this implementation of .NET and .NET Framework 4.8. The name "Core" was dropped to make clear that this is now the main implementation of .NET.

Reviewer: ```suggestion
In 2014, Microsoft began writing a cross-platform, open-source successor to .NET Framework. This new implementation of .NET was named .NET Core until it reached version 3.1. The next version after .NET Core 3.1 is .NET 5.0, which will be released in November 2020. Version number 4 was skipped to avoid confusion between this implementation of .NET and .NET Framework 4.8. The name "Core" was dropped to make clear that this is now the main implementation of .NET.
```

  |---------|---------------------|
  | .NET Framework | The original, Windows-only implementation of .NET. "Framework" is capitalized. |
  | target framework | The collection of APIs that a .NET app or library relies on. Examples: .NET Core 3.1, .NET Standard 2.0 |
  | Target Framework Moniker (TFM)  | A TFM is a standardized token format for specifying the target framework of a .NET app or library. Example: `net462` for .NET Framework 4.6.2. |
  | framework-dependent app | An app that can only run on a machine where you've installed the runtime from the [.NET download page](https://dotnet.microsoft.com/download/dotnet-core). "Framework" in this usage is the same thing as the "runtime" that you download from the .NET download page. |
  
* **platform**

  |Context  | "platform" meaning |
  |---------|--------------------|
  | cross platform | In this term, "platform" means an operating system and the hardware it runs on, such as Windows, macOS, Linux, iOS, and Android. |
  | .NET platform | Usage varies. The reference may be to one implementation of .NET (such as .NET Framework or .NET 5) or to an overarching concept of .NET including all implementations. |

> 

Reviewer: I think that SDK could also be added to the overloaded list.

Author: Good point.  Contexts and meanings would be (1) Download page: a collection of app development tools, (2) Project file -- a set of MSBuild targets and tasks for a given app type. Are there others?

Reviewer: That sounds great to me. I can't think of others.

Author: I added SDK.  Thanks for the review!

For more information about .NET terminology, see the [.NET glossary](../standard/glossary.md).

## Advanced scenarios

The following sections explain some capabilities of .NET that are useful in advanced scenarios.

### Native interop

Every operating system includes an application programming interface (API) that provides system services. .NET provides several ways to call those APIs.

> The main way to do native interoperability is via "platform invoke" or P/Invoke for short. P/Invoke is supported across Linux and Windows platforms. A Windows-only way of doing native interoperability is known as "COM interop," which is used to work with [COM components](/cpp/atl/introduction-to-com) in managed code. It's built on top of the P/Invoke infrastructure, but it works in subtly different ways.

Reviewer: ```suggestion
The main way to interoperate with native APIs is via "platform invoke" or P/Invoke for short. P/Invoke is supported across Linux and Windows platforms. A Windows-only way of interoperating is known as "COM interop," which works with [COM components](/cpp/atl/introduction-to-com) in managed code. It's built on top of the P/Invoke infrastructure, but it works in subtly different ways.
```

For more information, see [Native interoperability](../standard/native-interop/index.md).

### Unsafe code

> Depending on language support, the CLR lets you access native memory and do pointer arithmetic via `unsafe` code. These operations are needed for certain algorithms and system interoperability. Although powerful, use of unsafe code is discouraged unless it's necessary to interop with system APIs or implement the most efficient algorithm. Unsafe code may not execute the same way in different environments and also loses the benefits of a garbage collector and type safety. It's recommended to confine and centralize unsafe code as much as possible and test that code thoroughly.

Reviewer: ```suggestion
Depending on language support, the CLR lets you access native memory and do pointer arithmetic via `unsafe` code. These operations are needed for certain algorithms and system interoperability. Although powerful, use of unsafe code is discouraged unless it's necessary to interoperate with system APIs or implement the most efficient algorithm. Unsafe code may not execute the same way in different environments and also loses the benefits of a garbage collector and type safety. It's recommended to confine and centralize unsafe code as much as possible and test that code thoroughly.
```

---
title: .NET introduction and overview
description: Learn about .NET, a free, open-source development platform for building many kinds of apps.
author: tdykstra
ms.date: 09/28/2020
ms.custom: "updateeachrelease"
---
# Introduction to .NET

.NET is a free, open-source development platform for building many kinds of apps, such as:

> * [Web apps, web APIs, and microservices](/aspnet/core/introduction-to-aspnet-core#recommended-learning-path)
> * [Serverless functions in the cloud](/azure/azure-functions/functions-create-first-function-vs-code?pivots=programming-language-csharp)
> * [Mobile apps](https://dotnet.microsoft.com/learn/xamarin/hello-world-tutorial/intro)
> * Desktop apps
>   * [Windows WPF](/dotnet/desktop/wpf/)
>   * [Windows Forms](/dotnet/desktop/winforms/)
>   * [Universal Windows Platform (UWP)](/windows/uwp/get-started/create-a-hello-world-app-xaml-universal)
> * [Games](https://dotnet.microsoft.com/learn/games/unity-tutorial/intro)
> * [Internet of Things (IoT)](https://dotnet.microsoft.com/apps/iot)
> * [Machine learning](../machine-learning/index.yml)
> * [Console apps](tutorials/with-visual-studio-code.md)
> * [Windows services](/aspnet/core/host-and-deploy/windows-service)

Reviewer 2: ```suggestion
* [Web apps, web APIs, and microservices](/aspnet/core/introduction-to-aspnet-core#recommended-learning-path)
* [Serverless functions in the cloud](/azure/azure-functions/functions-create-first-function-vs-code?pivots=programming-language-csharp)
* [Cloud native apps](/dotnet/architecture/cloud-native/)
* [Mobile apps](https://dotnet.microsoft.com/learn/xamarin/hello-world-tutorial/intro)
* Desktop apps
  * [Windows WPF](/dotnet/desktop/wpf/)
  * [Windows Forms](/dotnet/desktop/winforms/)
  * [Universal Windows Platform (UWP)](/windows/uwp/get-started/create-a-hello-world-app-xaml-universal)
* [Games](https://dotnet.microsoft.com/learn/games/unity-tutorial/intro)
* [Internet of Things (IoT)](https://dotnet.microsoft.com/apps/iot)
* [Machine learning](../machine-learning/index.yml)
* [Console apps](tutorials/with-visual-studio-code.md)
* [Windows services](/aspnet/core/host-and-deploy/windows-service)
```

Added cloud native apps.
